accepted/2021/json-polymorphism.md

```diff
@@ -0,0 +1,275 @@
+the serializer will apply the JSON schema derived from the type `Foo`:
+```csharp
+Foo foo1 = new Foo { A = 1 };
+Foo foo2 = new Bar { A = 1, B = 2 };
+Foo foo3 = new Baz { A = 1, B = 2, C = 3 };
+
+JsonSerializer.Serialize<Foo>(foo1); // { "A" : 1 }
+JsonSerializer.Serialize<Foo>(foo2); // { "A" : 1 }
+JsonSerializer.Serialize<Foo>(foo3); // { "A" : 1 }
+```
+Under the new proposal we can change this behaviour by annotating 
+the base class (or interface) with the `JsonPolymorphicType` attribute:
+```csharp
+[JsonPolymorhicType]
+public class Foo
+{
+  ...
+}
+```
+which will result in the above values now being serialized as follows:
+```csharp
+JsonSerializer.Serialize<Foo>(foo1); // { "A" : 1 }
+JsonSerializer.Serialize<Foo>(foo1); // { "A" : 1, "B" : 2 }
+JsonSerializer.Serialize<Foo>(foo2); // { "A" : 1, "B" : 2, "C" : 3 }
+```
+Note that the `JsonPolymorphic` attribute is not inherited by derived types.
+In the above example `Bar` inherits from `Foo` yet is not polymorphic in its own right:
+```csharp
+Bar bar = new Baz { A = 1, B = 2, C = 3 };
+JsonSerializer.Serialize<Bar>(bar); // { "A" : 1, "B" : 2 }
+```
+If annotating the base class with an attribute is not possible,
+polymorphism can alternatively be opted in for a type using the
+new `JsonSerializerOptions.SupportedPolymorphicTypes` predicate:
+```csharp
+public class JsonSerializerOptions
+{
+  public Func<Type, bool> SupportedPolymorphicTypes { get; set; }
+}
+```
+Applied to the example above:
+```csharp
+var options = new JsonSerializerOptions { SupportedPolymorphicTypes = type => type == typeof(Foo) };
+JsonSerializer.Serialize<Foo>(foo1, options); // { "A" : 1, "B" : 2 }
+JsonSerializer.Serialize<Foo>(foo2, options); // { "A" : 1, "B" : 2, "C" : 3 }
+```
+It is always possible to use this setting to enable polymorphism _for every_ serialized type:
+```csharp
+var options = new JsonSerializerOptions { SupportedPolymorphicTypes = _ => true };
+
+// `options` treats both `Foo` and `Bar` members as polymorphic
+Baz baz = new Baz { A = 1, B = 2, C = 3 };
+JsonSerializer.Serialize<Foo>(baz, options); // { "A" : 1, "B" : 2, "C" : 3 }
+JsonSerializer.Serialize<Bar>(baz, options); // { "A" : 1, "B" : 2, "C" : 3 }
+```
+As mentioned previously, this feature provides no provision for deserialization.
+If deserialization is a requirement, users would need to opt for the
+polymorphic serialization with type discriminators feature.
+
+## Polymorphism with type discriminators
+
+This feature allows users to opt in to polymorphic serialization for a given type
+by associating string identifiers with particular subtypes in the hierarchy.
+These identifiers are written to the wire so this brand of polymorphism is roundtrippable.
+
+At the core of the design is the introduction of `JsonKnownType` attribute that can
+be applied to type hierarchies like so:
+```csharp
+[JsonKnownType(typeof(Derived1), "derived1")]
+[JsonKnownType(typeof(Derived2), "derived2")]
```

Comment:
Ditto, could a user want to ignore these attributes for non-owned types? Is that a valid scenario?

In this case it looks like they could provide a `TypeDiscriminatorConfiguration` for the type with no known types... assuming we like that behavior and runtime `JsonSerializerOptions` win over design-time attributes.

Reply:
Similarly, one could specify a custom converter to opt out from these attributes. Using an empty `TypeDiscriminatorConfiguration` instance would be a valid alternative, provided we opt for "lax mode" semantics as described in the "open questions" section.

Reply:
👍 https://github.com/dotnet/designs/pull/226#discussion_r650055156